waku/waku_sync/protocol.nim

```diff
@@ -0,0 +1,475 @@
+      payload = reconciled.payload
+
+    let received = ?await sent.listenBack()
+
+    trace "sync payload received",
+      client = self.peerManager.switch.peerInfo.peerId,
+      server = conn.peerId,
+      payload = received.payload
+
+    reconciled = (?received.clientReconcile(hashes)).valueOr:
+      let completed = error # Result[Reconciled, Completed]
+
+      ?await completed.clientTerminate()
+
+      debug "sync session ended gracefully",
+        client = self.peerManager.switch.peerInfo.peerId, server = conn.peerId
+
+      return ok(hashes)
+
+proc sync*(
+    self: WakuSync, peerInfo: Option[RemotePeerInfo] = none(RemotePeerInfo)
+): Future[Result[(seq[WakuMessageHash], RemotePeerInfo), string]] {.async, gcsafe.} =
+  let peer =
+    if peerInfo.isSome():
+      peerInfo.get()
+    else:
+      let peer: RemotePeerInfo = self.peerManager.selectPeer(WakuSyncCodec).valueOr:
+        return err("No suitable peer found for sync")
+
+      peer
+
+  let connOpt = await self.peerManager.dialPeer(peer, WakuSyncCodec)
+
+  let conn: Connection = connOpt.valueOr:
+    return err("Cannot establish sync connection")
+
+  let hashes: seq[WakuMessageHash] = (await self.request(conn)).valueOr:
+    debug "sync session ended",
+      server = self.peerManager.switch.peerInfo.peerId, client = conn.peerId, error
+
+    return err("Sync request error: " & error)
+
+  return ok((hashes, peer))
+
+proc handleLoop(
+    self: WakuSync, conn: Connection
+): Future[Result[seq[WakuMessageHash], string]] {.async.} =
+  let (start, `end`) = calculateRange(self.relayJitter)
+
+  let initialized =
+    ?serverInitialize(self.storage, conn, self.maxFrameSize, start, `end`)
+
+  var sent = initialized
+
+  while true:
+    let received = ?await sent.listenBack()
+
+    trace "sync payload received",
+      server = self.peerManager.switch.peerInfo.peerId,
+      client = conn.peerId,
+      payload = received.payload
+
+    let reconciled = (?received.serverReconcile()).valueOr:
+      let completed = error # Result[Reconciled, Completed]
+
+      let hashes = await completed.serverTerminate()
+
+      return ok(hashes)
+
+    sent = ?await reconciled.send()
+
+    trace "sync payload sent",
+      server = self.peerManager.switch.peerInfo.peerId,
+      client = conn.peerId,
+      payload = reconciled.payload
+
+proc initProtocolHandler(self: WakuSync) =
+  proc handle(conn: Connection, proto: string) {.async, gcsafe, closure.} =
+    debug "sync session requested",
+      server = self.peerManager.switch.peerInfo.peerId, client = conn.peerId
+
+    let hashes = (await self.handleLoop(conn)).valueOr:
+      debug "sync session ended",
+        server = self.peerManager.switch.peerInfo.peerId, client = conn.peerId, error
+
+      #TODO send error code and desc to client
+      return
+
+    if hashes.len > 0 and self.transferCallBack.isSome():
+      let callback = self.transferCallBack.get()
+
+      (await callback(hashes, conn.peerId)).isOkOr:
```

Comment:
Callbacks are :hankey: . Yes we can use private procs but `WakuSync` would have to reference `WakuArchive` and  `WakuStoreClient` which is also :hankey: 

It goes back to my discussion with @jm-clius IMO there's no way to abstract away archive and store client....

What flavor of :hankey: do we want?

Reply:
I feel it, but still it can be an option to move out these refs out of this protocol into a separate callback setup place. That would eliminate the need for refs of WakuArchive and WakuStoreClient.
Now they need to be provided with new*.
In other words by separating setting up the callback from the protocol will move away the decision how that is implemented. Also enables mocking it for testing. WDYT?

Reply:
> In other words by separating setting up the callback from the protocol will move away the decision how that is implemented.

The only other place would be in `wakunode` and that's not good either. I made some changes but IDK what to do.